### ISSUE TITLE
Support Client Streaming for fetch API

### ISSUE THREAD

**Issue by oott123** [connectrpc/connect-es#669]
**Is your feature request related to a problem? Please describe.**
We cannot use client stream on web.

**Describe the solution you'd like**
Support client stream requests with the fetch API by default.

**Describe alternatives you've considered**
I can write use own transport function.

**Additional context**
[Streaming requests with the fetch API](https://developer.chrome.com/articles/fetch-streaming-requests/) by Chrome Developers

**Comment by oott123** [comment]
Actually, Firefox already supported this feature, at least here on my machine, and yes I completely agree that we should add tests to make sure that it works correctly.

**Comment by timostamm** [comment]
> Actually, Firefox already supported this feature, at least here on my machine

Could you share which version you are running? We did not see the feature detection recommended by the Chrome Developers Blog working with the [developer edition](https://www.mozilla.org/en-CA/firefox/developer/).

**Comment by oott123** [comment]
@timostamm  I'm so sorry, I have re-checked this on my Firefox (114.0.1) browser and it do does NOT work. Sorry for my mistake, I may confused between browsers.

**Comment by smaye81** [comment]
If you want a pure Connect solution, the best option for large files would probably be to chunk up the file on the client and then send individual unary requests to the backend and then reassembling the chunks there.

In addition, depending on the size of the file, you could potentially send it all in one request rather than having to chunk it.

**Comment by StarpTech** [comment]
@smaye81 came to the same conclusion but since client-side streaming is not supported on the browser what's the workaround there?

**Comment by smaye81** [comment]
Since client-streaming is not yet supported from browsers, those are probably your options. You could use regular HTTP and multipart form data, also. But if you wanted a pure Connect approach from a browser, then your options are limited unfortunately.

Some gRPC-Web implementations such as Improbable's gRPC-Web client facilitate client-streaming via websockets, but Connect does not support that at the moment.

**Comment by StarpTech** [comment]
This is sad but understandable. It would be nice if this workaround is somehow managed by Connect protocol. As a Dev, I need to plumb all together again, which I wanted to prevent by using Connect in the first place.

**Comment by smaye81** [comment]
Yeah, that's fair. We're keeping an eye on support for sending `ReadableStream` and evaluating possible alternatives, so hopefully, we have an answer for this in the near future.

**Comment by hydrz** [comment]
#[b5e5c96](https://github.com/opensraph/connect-es/commit/b5e5c96b909905f1de3c0ef2966a48f221315ab0)

A compatibility solution is to use half-duplex requests on supporting browsers, or fall back to buffering the entire request stream before sending it. I've modified the gRPC-web transport as well, and it works properly now. Hope this helps!

**Comment by moozechen** [comment]
This would be really valuable for real-time collaboration use cases. We've been evaluating Connect vs raw WebSocket for a consulting platform that needs typed bidirectional streaming in the browser, and the lack of client streaming support in fetch API is the main blocker for pure browser implementations.

One workaround we've tested: using server-sent events (SSE) for the server→client stream and regular fetch POST for client→server messages. It's not true bidirectional streaming but covers ~80% of use cases (dashboards, notifications, log streaming) with zero library overhead since SSE is browser-native and works through all proxies/CDNs. For the 20% that truly need full-duplex (chat, collaborative editing), WebSocket with protobuf serialization gets you most of the type safety with more browser compatibility.

Curious if the team has considered an SSE-based transport as a progressive enhancement path before full fetch streaming lands?
